src/lib.rs

```diff
@@ -240,7 +237,41 @@
 //! [Serialize]: https://docs.rs/serde/1/serde/trait.Serialize.html
 //! [Set]: https://docs.rs/fixed-map/*/fixed_map/map/struct.Set.html
 
-#![deny(missing_docs)]
+#![forbid(unsafe_code)]
+// Enable more useful rustc lints
+#![deny(
+    absolute_paths_not_starting_with_crate,
+    elided_lifetimes_in_paths,
+    explicit_outlives_requirements,
+    keyword_idents,
+    macro_use_extern_crate,
+    meta_variable_misuse,
+    missing_copy_implementations,
+    missing_docs,
+    non_ascii_idents,
+    noop_method_call,
+    pointer_structural_match,
+    single_use_lifetimes,
+    trivial_casts,
+    trivial_numeric_casts,
+    unreachable_pub,
+    unused_extern_crates,
+    unused_import_braces,
+    unused_lifetimes,
+    unused_macro_rules,
+    unused_qualifications,
+    unused_tuple_struct_fields,
+    variant_size_differences
+)]
```

Comment:
Scary looking list!

Now I'm not entirely sure enabling `deny` on lints for library crates is future proof. If rustc adds future lint violations which falls under these (maybe something wasn't caught previously) wouldn't that cause a compilation error?

Could you make sure this is not the case? Otherwise I'm :+1:.

Reply:
So I don't believe this is guaranteed to be future proof unfortunately even if we stick to a concrete list: https://rust-unofficial.github.io/patterns/anti_patterns/deny-warnings.html

Now a better way to ensure that a specific crate avoids these now and into the future is to only enable them for the CI build, like through some feature flag.

Reply:
That anti-pattern is specifically about `deny(warnings)`, because they could set a given lint to warn by default (which can be done without an edition). It's not about changing existing lints. And it doesn't apply to clippy lints, because they aren't enabled unless you're running clippy.

If you're really worried about it, I would have just enabled them as warnings instead - that way you'll still see them in your editor unconditionally.

Regardless, you could have made it a little nicer by leaving the long list of deny lints in a single deny(). That would have left less `cfg_attr` noise.

Also, it there a reason you changed `forbid(unsafe)` to `deny(unsafe)`? Forbid had stronger guarantees - you can't re-allow it later on. Changing it to deny kind of defeats the purpose.

Reply:
> That anti-pattern is specifically about deny(warnings), because they could set a given lint to warn by default (which can be done without an edition). It's not about changing existing lints. And it doesn't apply to clippy lints, because they aren't enabled unless you're running clippy.

Specifically yes, but reading the full article should convey that it applies lints that might *change* in the future as well. I.e. it would only be appropriate to `deny` something that is very unlikely to ever change, like `missing_docs`. I'm not sure how it applies in the future to the rest of the lints but better safe than sorry. Catching them in CI (i.e. as they're introduced as a change in this crate) fulfills the same role of ensuring idiomatic code.

> If you're really worried about it, I would have just enabled them as warnings instead - that way you'll still see them in your editor unconditionally.

Yep, this sounds like a better option! Then you'd only need to add `-D warnings` in CI. If you feel like doing a follow up PR, feel free to!

> Regardless, you could have made it a little nicer by leaving the long list of deny lints in a single deny(). That would have left less `cfg_attr` noise.

Yeah, all though I prefer the `cfg_attr` noise in this instance ;).

> Also, it there a reason you changed `forbid(unsafe)` to `deny(unsafe)`? Forbid had stronger guarantees - you can't re-allow it later on. Changing it to deny kind of defeats the purpose.

Ah, no. That was a mistake during editing. What exactly is the difference between `forbid` and `deny`?

Reply:
> What exactly is the difference between forbid and deny?

`deny` can be overridden by an `allow` later on. `forbid` can not.

> The reason is that they can be quite noisy for downstream users if they're introduced by compiler upgrades.

Aren't warnings in dependencies hidden from downstream users?

Reply:
OK, if that's true then changing it to always warn and `-D warnings` in CI is fine! Cheers!